File: backend/app/agents/runner.py
import asyncio
import logging

from fastapi import FastAPI

from app.config import get_settings
from app.database import SessionLocal

from .supply_agent import SupplyAgent

logger = logging.getLogger(__name__)
agent = SupplyAgent()


async def _agent_loop():
    # Periodic loop: query DB and run agent checks on the configured interval
    interval = get_settings().agent_poll_interval_seconds
    while True:
        try:
            db = SessionLocal()
            try:
                num = agent.run_check(db)
                logger.info("Agent run completed, %d events", num)
            finally:
                db.close()
            await asyncio.sleep(interval)
        except asyncio.CancelledError:
            break
        except Exception:
            logger.exception("Agent loop failed")


def start_agents(app: FastAPI):
    task = asyncio.create_task(_agent_loop())
    app.state.agent_task = task
    return task

File: src/lib/server/agents/runner.ts
/** Periodic runner loop for the supply agent. */

import { getSettings } from "../config";
import { SupplyAgent } from "./supply";

const globalRunner = globalThis as typeof globalThis & {
  __nexusGridSupplyAgent?: SupplyAgent;
  __nexusGridAgentTimer?: NodeJS.Timeout;
};

/** The process-wide supply agent, so its memory survives hot reloads. */
export function getSupplyAgent(): SupplyAgent {
  if (!globalRunner.__nexusGridSupplyAgent) {
    globalRunner.__nexusGridSupplyAgent = new SupplyAgent();
  }
  return globalRunner.__nexusGridSupplyAgent;
}

/**
 * Start the periodic scan loop.
 *
 * Each tick runs the agent against current inventory on the configured
 * interval. The timer is unref'd so it never keeps the process alive on its
 * own, and is stored on `globalThis` so a hot reload replaces rather than
 * duplicates it.
 */
export function startAgents(): void {
  const settings = getSettings();
  if (settings.skipAgentStartup) {
    console.info("Agent startup skipped (SKIP_AGENT_STARTUP)");
    return;
  }

  if (globalRunner.__nexusGridAgentTimer) {
    clearInterval(globalRunner.__nexusGridAgentTimer);
  }

  let running = false;
  const timer = setInterval(async () => {
    // Skip a tick rather than overlapping if the previous scan is still going
    if (running) return;
    running = true;
    try {
      const events = await getSupplyAgent().runCheck();
      console.info(`Agent run completed, ${events} events`);
    } catch (error) {
      console.error("Agent loop failed", error);
    } finally {
      running = false;
    }
  }, settings.agentPollIntervalSeconds * 1000);

  timer.unref?.();
  globalRunner.__nexusGridAgentTimer = timer;
  console.info(`Agent loop started (every ${settings.agentPollIntervalSeconds}s)`);
}
